File: firefox/src/java/org/openqa/selenium/firefox/FirefoxLauncher.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium.firefox;

import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.firefox.internal.RunningInstanceConnection;

import java.io.IOException;
import java.net.ConnectException;

public class FirefoxLauncher {
  private final FirefoxBinary binary;

  public FirefoxLauncher(FirefoxBinary binary) {
      this.binary = binary;
  }

  public static void main(String[] args) {
    FirefoxBinary binary = new FirefoxBinary();
    FirefoxLauncher launcher = new FirefoxLauncher(new FirefoxBinary());

    String profileName = "WebDriver";
    int port = FirefoxDriver.DEFAULT_PORT;
    if (args.length >= 2) {
      port = Integer.parseInt(args[1]);
    }
    
    if (args.length >= 1) {
      profileName = args[0];
    }
    
    // If there's a browser already running, connect and kill it.
    launcher.connectAndKill(port);
    
    // Ensure the profile is created, and initialize it.
    launcher.createBaseWebDriverProfile(binary, profileName, port);

    // Connect until it works.
    launcher.repeatedlyConnectUntilFirefoxAppearsStable(port);
  }
    
  public FirefoxBinary startProfile(FirefoxProfile profile, int port) throws IOException {
    FirefoxBinary binaryToUse = binary;
    if (binary == null) {
      binaryToUse = new FirefoxBinary();
    }

    FirefoxProfile profileToUse = profile.createCopy(port);
    binaryToUse.clean(profileToUse);
    binaryToUse.startProfile(profileToUse);
    return binaryToUse;
  }
  
  @Deprecated
  public void createBaseWebDriverProfile(FirefoxBinary binary, String profileName, int port) {
    // If there's a browser already running
    connectAndKill(port);

    System.out.println(String.format("Creating %s", profileName));
    try {
        binary.createProfile(profileName);
        System.out.println("Profile created");
        binary.waitFor();
    } catch (IOException e) {
        throw new WebDriverException("Unable to create base webdriver profile", e);
    } catch (InterruptedException e) {
        throw new WebDriverException(e);
    }
    
    ProfileManager.getInstance().createProfile(binary, profileName, port);
  }

  @Deprecated
  protected void connectAndKill(int port) {
    try {
        ExtensionConnection connection = new RunningInstanceConnection("localhost", port, 5000);
        connection.quit();
    } catch (ConnectException e) {
        // This is fine. It just means that Firefox isn't running with the webdriver extension installed already
    } catch (NotConnectedException e) {
        // This is fine. It just means that Firefox isn't running with the webdriver extension installed already
    } catch (IOException e) {
        throw new WebDriverException(e);
    }
  }

  private void repeatedlyConnectUntilFirefoxAppearsStable(int port) {
    ExtensionConnection connection;
    
    // maximum wait time is a minute
    long maxWaitTime = System.currentTimeMillis() + binary.getTimeout();
      
    do {
      try {
          connection = new RunningInstanceConnection("localhost", port, 1000);
          Thread.sleep(2000);
          connection.quit();
          return;
      } catch (ConnectException e) {
          // Fine. Nothing listening. Perhaps in a restart?
      } catch (IOException e) {
          // Expected. It'll do that
      } catch (InterruptedException e) {
          throw new WebDriverException(e);
      }
    } while (System.currentTimeMillis() < maxWaitTime);
  }
}

File: firefox/src/py/firefoxlauncher.py
# Copyright 2008-2009 WebDriver committers
# Copyright 2008-2009 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Launches the firefox and does necessary preparation like
installing the extension"""


from subprocess import Popen
import logging
import time
import os
from extensionconnection import ExtensionConnection
import utils

MAX_START_ATTEMPTS = 60

class FirefoxLauncher(object):
    """Launches the firefox browser."""

    def __init__(self):
        self.extension_connection = ExtensionConnection()
        self._start_cmd = utils.get_firefox_start_cmd()
        self.process = None

    def launch_browser(self, profile):
        """Launches the browser for the given profile name.
        It is assumed the profile already exists.
        """
        self.profile = profile
        while self.extension_connection.is_connectable():
            logging.info("Browser already running, kill it")
            self.extension_connection.connect_and_quit()
            time.sleep(1)
        self._start_from_profile_path(profile.path)

        attempts = 0
        while not self.extension_connection.is_connectable():
            attempts += 1
            if attempts >  MAX_START_ATTEMPTS:
                raise RuntimeError("Unablet to start firefox")
            self._start_from_profile_path(profile.path)
            time.sleep(1)

    def _lock_file_exists(self):
        return os.path.exists(os.path.join(self.profile.path, ".parentlock"))

    def kill(self):
        """Kill the browser.

        This is useful when the browser is stuck.
        """
        try:
            if self.process:
                os.kill(self.process.pid, 9)
        except AttributeError:
            # kill may not be available under windows environment
            pass

    def _start_from_profile_path(self, path):
        os.environ["XRE_PROFILE_PATH"] = path
        self.process = Popen([self._start_cmd, "-no-remote", "--verbose"])

    def _wait_until_connectable(self):
        """Blocks until the extension is connectable in the firefox."""
        while not self.extension_connection.is_connectable():
            logging.debug("Waiting for browser to launch...")
            if self.process.returncode:
                # Browser has exited
                return False
            time.sleep(1)
        return True
